rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,358 @@
+uploaded to S3. The [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)
+will be used when uploading to S3 to verify the integrity of each chunk. The AWS SDK will be useful in simplifying the
+upload to S3. For calculating the checksum while streaming, [checksumming_io](https://github.com/HumanCellAtlas/checksumming_io)
+can be used.
+
+Once the upload is complete, the download job will be removed from the queue. If the final hash does not match, then
+the upload process will be restarted. The retry logic will be an exponential backoff, with a max retry of 5 times.
+
+While streaming the file, the compute resource will periodically check if the upload has been canceled using the upload
+table. If the upload status is cancel pending, the compute resource will stop the stream, delete any data that has been uploaded,
+mark the job as complete in the queue, and set the status of the upload to cancelled.
+
+#### 7. Retry
+
+If the upload fails for any reason, the compute resource will mark the status of the upload as waiting, and return the upload
+job to the queue.
+
+#### 8. Failed Uploads
+
+If an upload fails enough times, it will be removed from the download queue. The upload status will change to failed.
+
+### Components
+
+This section describes the purpose of components that make up the upload design.
+
+#### Data Portal Browser App
+
+The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
+It will be used to authenticate the user and interact with the DP Backend to upload files.
+
+#### Data Portal Backend
+
+The DP Backend is responsible for handling API requests. It will enqueue shared links for downloading, and update the
+upload status. This additional functionality can be built into the existing DP Backend Lambda.
+
+The backend will also be responsible for performing a simple validation of the file being uploaded. The validation is
+performed during the initial request to upload. The validation consists of verifying the extension of the file is
+".h5ad".
+
+#### Download Queue
+
+The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending download jobs.
+The download jobs are used by the upload service to know where to download the file from before uploading them to S3.
+
+##### Download Job Entry
+
+These are the fields that will be in the download job placed in the queue:
+
+- Link - the link to download. Use to retrieve the file from CSP.
+- submission_id - identifies the submission. Used to determine the storage location.
+- dataset_id - identifies the dataset. Used to determine the storage location
+- file_name - the name of the file being downloaded
+
+#### Upload Service
+
+A computing resource within the AWS cloud. It will receive download jobs from the Download Queue. Those jobs contain
```

Comment:
Instead of "It will receive download jobs", maybe "It will poll for download jobs"

Reply:
Is there a mechanism to increase the number of Upload Services in the case where we have more than one upload at a time?

Reply:
We can setup a cloudwatch alert if the length of SQS messages gets too long. The alert can trigger additional compute resources to spawn.